### ISSUE TITLE
Allow to work with local actions outside of the repository we run the workflows from

### ISSUE THREAD

**Issue by Porkepix** [nektos/act#1579]
### Act version

act version 0.2.34

### Feature description

It would be a nice feature, when developing a new action or more, several actions (for example to migrate from CircleCI's orbs to GitHub's actions) to be able to have like a "path" (I mean, the same way you'd add a location to your $PATH for your shell) act would be looking at/mounting, so that when fetching actions it could also get them from here, to allow for actions development offline and not only the workflows themselves.

Note that I might have missed the way to do so, but the best I could find was to copy and paste actions within the repo itself and change to call from the `org/repo` syntax to a local one

**Comment by trash-anger** [comment]
I was about to create quite a similar feature request!

I'm working in a company which have quite a flexible remote work policy leading developers to travel a lot by flight. (which is currently my case)

I'm a new user of act and when I tried to use act during my previous flight I just got this message:
`FATA[0000] dial udp 8.8.8.8:80: connect: network is unreachable`

So landing I've been wondering how we could work offline an easy way, it could be nice to implement a caching solution allowing to fetch pipeline dependencies while running it. Then running it again would first try to reach online first (and update the cache as a result) and offline, would rely on the cache.

I'm currently adapting the pipelines of my company to get a custom behaviour depending on the `GITHUB_ACTIONS` var to allow anyone to build locally, and deploy the current component to a local k8s cluster.

I'm thinking out loud:
> Another interesting thing would be to implement a similar feature than the ".github" repository. I haven't check if it's possible to make it work with a simlink added to the gitignore tho.

@Porkepix , maybe you could use the workflow reuse function if it works and if you can refactor your pipeline.

**Comment by Porkepix** [comment]
> I was about to create quite a similar feature request!
> 
> I'm working in a company which have quite a flexible remote work policy leading developers to travel a lot by flight. (which is currently my case)
> 
> I'm a new user of act and when I tried to use act during my previous flight I just got this message: `FATA[0000] dial udp 8.8.8.8:80: connect: network is unreachable`
> 
> So landing I've been wondering how we could work offline an easy way, it could be nice to implement a caching solution allowing to fetch pipeline dependencies while running it. Then running it again would first try to reach online first (and update the cache as a result) and offline, would rely on the cache.

Act actually already cache actions, see `~/.cache/act`. The issue here is rather that it tries to reach out Google's Public DNS which is curious to say the least. I noticed it too but didn't opened an issue, would definitely be interesting to know why it does that. And if every resources are available locally then, yes, I believe it should work 100% offline.

What I had in mind was rather to have a clone of the actions repos and to provide them to act like you would with libs given to a compiler, or bineries in your `$PATH`: a local place to look for those actions.

> @Porkepix , maybe you could use the workflow reuse function if it works and if you can refactor your pipeline.

I don't understand what you had in mind here.

**Comment by trash-anger** [comment]
I'm boarding, I'll do that tonight!

**Comment by cardinalby** [comment]
Act is great for testing Workflows. Developing individual actions I also tried to use act, but found that it's better to have integration tests. Especially for JavaScript actions where you can run your action locally offline, pass intputs, env, etc. and **debug** it, test its outputs, examine all commands issued by the action, etc. 

I use [github-action-ts-run-api](https://github.com/cardinalby/github-action-ts-run-api) for that.

**Comment by mcascone** [comment]
All I want is to be able to develop my actions locally, separate from the repo with the workflow file that calls them. 

Specifically, I want to be able to use a local path in the `uses:` part of a workflow:

## Existing behavior
```yaml
jobs:
  Deploy:
    steps:
      - uses: 'myOrg/my-action@my-branch'
```

## Desired functionality:
```yaml
jobs:
  Deploy:
    steps:
      - uses: '/path/to/local/my-action@my-branch'
```

## 
Going by the output, there would need to be a trigger or flag to not tack the `https://github.com` onto the path:
```sh
 git clone 'https://github.com/~/myOrg' # ref=my-branch
[My-Action/Deploy] Unable to clone https://github.com/~/myOrg refs/heads/my-branch: repository not found
```

**Comment by mverkerk-godaddy** [comment]
I'm struggling with this too - I'm building a library of actions that can be shared across repos within our github org and when I'm testing this library locally with ACT, the `uses: ` pieces should point to `.[/path]` where otherwise, they should point to `{org}/{repo}[/path]@ref` ...

Currently, I have to (temporarily) update everything from `{org}/{repo}[/path]@ref` to `.[/path]`, run ACT, and change everything back before I commit/push (easy to forget). I have written a bash script that takes care of this but that's not ideal. I'll include the script at the bottom of this message...

In my testing, I have tried setting an ENV variable and/or using `env.ACT` but unfortunately `uses: ` does not support variables (as far as I know) 

This example:

```
- uses: ${{ env.ACT && './.github/actions/my-action' || 'org/repo/.github/actions/my-action@main' }}
  with:
    some-input: input
```

... failed with this error:

`Error: Expected format {org}/{repo}[/path]@ref. Actual '${{ env.ACT && './.github/actions/my-action' || 'my-org/my-repo/.github/actions/my-action@main' }}' Input string was not in a correct format`

would be great if there was a flag like `--replace-ghe-path-with-local {org}/{repo}` that would make this switch for us!

btw: thanks for all the great work! really enjoying this tool! 👍🏼

Here's my `act_path_replace.sh` script:

```bash
#!/usr/bin/env bash
target=${1}

if [ "$target" == "--to-remote" ]; then
  echo "Updating [action.yml] files to point to [org/repo/.github/actions]"
  FILES=$(find . -name "action.yml" -type f -exec grep -l "uses:.*\.\/\.github\/actions\/" {} \;)
  for file in $FILES; do
    sed -i 's|./.github/actions|org/repo/.github/actions|g' "$file"
    sed -i '/uses:.*\.github\/actions\//!b;/@v2/b;s/$/@v2/' "$file"
  done
elif [ "$target" == "--to-local" ]; then
  echo "Updating [action.yml] files to point to [./.github/actions]"
  FILES=$(find . -name "action.yml" -type f -exec grep -l "uses:.*org\/repo\/\.github\/actions\/" {} \;)
  for file in $FILES; do
    sed -i 's|org/repo|.|g' "$file"
    sed -i 's|@v2$||g' "$file"
  done
else
  echo "Please specify a target: local or remote"
  exit 1
fi

```

**Comment by reneleonhardt** [comment]
> Please test this version of act [ChristopherHX#34](https://github.com/ChristopherHX/act/pull/34).

Why was it closed after #1954 has been merged? 🤔

**Comment by mcascone** [comment]
is there any documentation on this other than the help text?

```
 --local-repository stringArray                      

Replaces the specified repository and ref with a local folder 

e.g. https://github.com/test/test@v0=/home/act/test or test/test@v0=/home/act/test, 
the latter matches any hosts or protocols
```

**Comment by reneleonhardt** [comment]
@Porkepix Is your use case now possible?

I can't see from this --local-repository example if mcascone's intuitive feature request https://github.com/nektos/act/issues/1579#issuecomment-1574290406 works when replacing $PWD by the actual absolute path to the local `test/test` repo... if it's outside of $PWD 😅

**Comment by prvg-sso** [comment]
What if `act` could dynamically replace parts of yml before execution of them. One could specify in a separate file that when encountering a `uses: enterprise/repo-stuff@1.0.0` it should be replace with `uses: testenv/repo-stuff@1.2.0`. In that way one could easily swap an action (that might be used in multiple places) to another one. I don't like the idea to add `if: ${{ !env.ACT }}` directives to handle this.

**Comment by jsoref** [comment]
I was going to file a new issue for https://github.com/check-spelling-sandbox/nektos-act-issue-up-dir/actions/runs/11671555945 not working in act, but this issue _technically_ is the right issue.

Summary:
1. The GitHub Actions Runtime **does** allow paths outside the checkout directory, specifically using a syntax of `./../` (and presumably more than that, but whatever).
2. [github/codeql-action](https://github.com/github/codeql-action/) [heavily uses this feature](https://github.com/search?q=repo%3Agithub%2Fcodeql-action%20.%2F..%2F&type=code).
3. For check-spelling, instead of using that feature (because I didn't find it), I'm leaving my [action shims within `.git/`](https://github.com/search?q=repo%3Acheck-spelling%2Fcheck-spelling%20check-spelling-actions%20&type=code), but it would be much cleaner if I could leave them in `./../` instead.

[`.github/workflows/test.yml`](https://github.com/check-spelling-sandbox/nektos-act-issue-up-dir/blob/4e633baa77d08a18f3dcbf56f420e7279ac83e93/.github/workflows/test.yml):
```yaml
on:
  push:
jobs:
  test:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v4
    - run: |
        mkdir ../action
        mv * ../action
    - uses: ./../action
```

[`action.yml`](https://github.com/check-spelling-sandbox/nektos-act-issue-up-dir/blob/4e633baa77d08a18f3dcbf56f420e7279ac83e93/action.yml)
```yml
runs:
  using: node20
  main: main.js
```

[`main.js`](https://github.com/check-spelling-sandbox/nektos-act-issue-up-dir/blob/4e633baa77d08a18f3dcbf56f420e7279ac83e93/main.js)
```js
console.log("hello world")
```

# [GitHub](https://github.com/check-spelling-sandbox/nektos-act-issue-up-dir/actions/runs/11671555945/job/32498209442#step:4:3)
```shell
Run mkdir ../action
Run ./../action
hello world
```

# Act
```shell
[test.yml/test] ⭐ Run Main actions/checkout@v4
[test.yml/test]   🐳  docker cp src=/Users/jsoref/code/nektos/issues/up-dir/. dst=/Users/jsoref/code/nektos/issues/up-dir
[test.yml/test]   ✅  Success - Main actions/checkout@v4
[test.yml/test] ⭐ Run Main mkdir ../action
mv * ../action
[test.yml/test]   🐳  docker exec cmd=[bash -e /var/run/act/workflow/1] user= workdir=
[test.yml/test]   ✅  Success - Main mkdir ../action
mv * ../action
[test.yml/test] ⭐ Run Main ./../action
[test.yml/test]   ❌  Failure - Main ./../action
[test.yml/test] failed to read 'action.yml' from action './../action' with path '' of step file does not exist
failed to read 'action.yaml' from action './../action' with path '' of step file does not exist
failed to read 'Dockerfile' from action './../action' with path '' of step file does not exist
[test.yml/test] 🏁  Job failed
Error: Job 'test' failed
```
